src/display/touch_manager.js

```diff
@@ -0,0 +1,177 @@
+/* Copyright 2024 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+class TouchManager {
+  #container;
+
+  #isPinching = false;
+
+  #overlayManager;
+
+  #onPinching;
+
+  #onPinchEnd;
+
+  #signal;
+
+  #touchInfo = null;
+
+  #touchManagerAC;
+
+  #touchMoveAC = null;
+
+  static #MIN_TOUCH_DISTANCE_TO_PINCH = 0;
+
+  constructor({ container, overlayManager, onPinching, onPinchEnd, signal }) {
+    this.#container = container;
+    this.#overlayManager = overlayManager;
+    this.#onPinching = onPinching;
+    this.#onPinchEnd = onPinchEnd;
+    this.#touchManagerAC = new AbortController();
+    this.#signal = AbortSignal.any([signal, this.#touchManagerAC.signal]);
+
+    container.addEventListener("touchstart", this.#onTouchStart.bind(this), {
+      passive: false,
+      signal: this.#signal,
+    });
+
+    // The 35 is coming from:
+    //  https://searchfox.org/mozilla-central/source/gfx/layers/apz/src/GestureEventListener.cpp#36
+    //
+    // The properties TouchEvent::screenX/Y are in screen CSS pixels:
+    //  https://developer.mozilla.org/en-US/docs/Web/API/Touch/screenX#examples
+    // MIN_TOUCH_DISTANCE_TO_PINCH is in CSS pixels.
+    TouchManager.#MIN_TOUCH_DISTANCE_TO_PINCH ??=
+      35 / (window.devicePixelRatio || 1);
+  }
+
+  #onTouchStart(evt) {
+    if (!this.#touchMoveAC) {
+      this.#touchMoveAC = new AbortController();
+      const signal = AbortSignal.any([this.#signal, this.#touchMoveAC.signal]);
+      const container = this.#container;
+      const opt = { signal, passive: false };
+      container.addEventListener(
+        "touchmove",
+        this.#onTouchMove.bind(this),
+        opt
+      );
+      container.addEventListener("touchend", this.#onTouchEnd.bind(this), opt);
+      container.addEventListener(
+        "touchcancel",
+        this.#onTouchEnd.bind(this),
+        opt
+      );
+    }
+
+    if (evt.touches.length < 2) {
```

Comment:
This has lost the PresentationMode check, note https://github.com/mozilla/pdf.js/blob/fbac4ab5f35dd4556069b69af51f78dbaaa62229/web/app.js#L2641-L2644